system-addon/lib/TelemetrySender.jsm

```diff
@@ -0,0 +1,126 @@
+/* globals Preferences, Services, XPCOMUtils */
+
+const {interfaces: Ci, utils: Cu} = Components;
+
+Cu.import("resource://gre/modules/Preferences.jsm");
+Cu.importGlobalProperties(["fetch"]);
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+
+const ENDPOINT_PREF = "telemetry.ping.endpoint";
+const TELEMETRY_PREF = "telemetry";
+const LOGGING_PREF = "performance.log";
+
+const ACTION_NOTIF = "user-action-event";
+const PERFORMANCE_NOTIF = "performance-event";
+const COMPLETE_NOTIF = "tab-session-complete";
+const UNDESIRED_NOTIF = "undesired-event";
+
+// This is intentionally a different pref-branch than the SDK-based add-on
+// used, to avoid extra weirdness for people who happen to have the SDK-based
+// installed.  Though maybe we should just forcibly disable the old add-on?
+const PREF_BRANCH = "browser.newtabpage.activity-stream";
+
+function TelemetrySender(args) {
+  let prefArgs = {branch: PREF_BRANCH};
+  if (args) {
+    if ("prefInitHook" in args) {
+      prefArgs.initHook = args.prefInitHook;
+    }
+  }
+
+  this._prefs = new Preferences(prefArgs);
+
+  this.enabled = this._prefs.get(TELEMETRY_PREF);
+  this._onTelemetryPrefChange = this._onTelemetryPrefChange.bind(this);
+  this._prefs.observe(TELEMETRY_PREF, this._onTelemetryPrefChange);
+
+  this.logging = this._prefs.get(LOGGING_PREF);
+  this._onLoggingPrefChange = this._onLoggingPrefChange.bind(this);
+  this._prefs.observe(LOGGING_PREF, this._onLoggingPrefChange);
+
+  this._pingEndpoint = this._prefs.get(ENDPOINT_PREF);
+  this._onEndpointPrefChange = this._onEndpointPrefChange.bind(this);
+  this._prefs.observe(ENDPOINT_PREF, this._onEndpointPrefChange);
```

Comment:
Why do we even want to watch for endpoint changes? Isn't that not a great idea, if we leave the endpoint open to a pref then in theory someone could change it and re-route where their data is going... if we set it once it should be enough to keep it that way. Besides, you've got the fake prefs for testing so there really isn't a need to watch for this pref change
I vote to remove this and just set it once in the constructor 🙋